main.c

```diff
@@ -59,28 +59,63 @@ static void pointer_handle_leave(void *data, struct wl_pointer *wl_pointer,
 }
 
 static void seat_set_outputs_dirty(struct slurp_seat *seat) {
-	struct slurp_box box;
-	seat_get_box(seat, &box);
 	struct slurp_output *output;
 	wl_list_for_each(output, &seat->state->outputs, link) {
-		if (box_intersect(&output->logical_geometry, &box)) {
+		if (box_intersect(&output->logical_geometry, &seat->selection)) {
 			set_output_dirty(output);
 		}
 	}
 }
 
+static bool in_box(const struct slurp_box *box, int32_t x, int32_t y) {
+	return (box->x <= x
+					&& box->x + box->width >= x
+					&& box->y <= y
+					&& box->y + box->height >= y);
+}
+static int32_t box_size(const struct slurp_box *box) {
+	return box->width * box->height;
+}
+
+
+static int min(int a, int b) {
+	return (a < b) ? a : b;
+}
 static void pointer_handle_motion(void *data, struct wl_pointer *wl_pointer,
 		uint32_t time, wl_fixed_t surface_x, wl_fixed_t surface_y) {
 	struct slurp_seat *seat = data;
 	// the places the cursor moved away from are also dirty
-	if (seat->button_state == WL_POINTER_BUTTON_STATE_PRESSED) {
+	if (seat->has_selection) {
 		seat_set_outputs_dirty(seat);
 	}
 
 	seat->x = wl_fixed_to_int(surface_x) + seat->current_output->logical_geometry.x;
 	seat->y = wl_fixed_to_int(surface_y) + seat->current_output->logical_geometry.y;
-
+	if (seat->button_state == WL_POINTER_BUTTON_STATE_RELEASED) {
+		// find smallest box intersecting the cursor
+		seat->has_selection = false;
+		struct slurp_box_ll *ptr = seat->state->boxes;
+		while(ptr != NULL) {
+			if (in_box(&ptr->b, seat->x, seat->y)) {
+				if (seat->has_selection && box_size(&seat->selection) < box_size(&ptr->b)) {
```

Comment:
What about inverting the selection and putting the two lines below in the `if` instead of using a `goto`?